### PR TITLE
[SYCL][Graph] Add a shortcut for adding leaves as dependencies

### PR DESCRIPTION
**Pull request by mfrancepillois**
Adds a node property that allows users to easily add all leaves of a graph as dependencies when creating a node with the explicit API. 
Updates the spec with this new feature.
Adds unitests that check this behaviour.

### CODE REVIEW

**Comment by mfrancepillois** [comment]
Upstream PR https://github.com/intel/llvm/pull/11855

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 1]
```diff
@@ -336,6 +336,11 @@ class depends_on {
     depends_on(NodeTN... nodes);
 };
 
+class depends_on_all_leaves {
+public:
+  depends_on_all_leaves() = default;
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:341]: nitpick, the indentation of  `public:` and the constructor isn't the same as the definitions above

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 2]
```diff
@@ -479,6 +484,20 @@ class depends_on {
...
+The API for explicitly adding nodes to a `command_graph` includes a
+`property_list` parameter. This extension defines the `depends_on_all_leaves` 
+property to be passed here. `depends_on_all_leaves` provides a shortcut for 
+adding all the current leaves of a graph as dependencies.
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental::property::node {
+class depends_on_all_leaves {
+public:
+  depends_on_all_leaves();
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:497]: same indentation comment

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 3]
```diff
@@ -479,6 +484,20 @@ class depends_on {
...
 ----
 
+==== Depends-On-All-Leaves Property 
+The API for explicitly adding nodes to a `command_graph` includes a
+`property_list` parameter. This extension defines the `depends_on_all_leaves` 
+property to be passed here. `depends_on_all_leaves` provides a shortcut for 
+adding all the current leaves of a graph as dependencies.
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental::property::node {
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:494]: Think we need a `}` on the last line of this code snippet to close the namespace, like in the snippet above

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 4]
```diff
@@ -239,6 +263,11 @@ class __SYCL_EXPORT modifiable_command_graph {
   /// @return Node added to the graph.
   node addImpl(const std::vector<node> &Dep);
 
+  /// Adds all graph leaves as dependencies
+  /// @param Node destination node to which the leaves of the graph will be
+  /// added as dependencies
```
Comment (EwanC) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp:268]:
```suggestion
  /// @param Node Destination node to which the leaves of the graph will be
  /// added as dependencies.
```

**Review by EwanC — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 5]
```diff
@@ -171,9 +187,17 @@ class __SYCL_EXPORT modifiable_command_graph {
...
-      return addImpl(CGF, Deps.get_dependencies());
+      node Node = addImpl(CGF, Deps.get_dependencies());
+      if (PropList.has_property<property::node::depends_on_all_leaves>()) {
+        addGraphLeafDependencies(Node);
+      }
+      return Node;
+    }
+    node Node = addImpl(CGF, {});
+    if (PropList.has_property<property::node::depends_on_all_leaves>()) {
+      addGraphLeafDependencies(Node);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp:198]: Should we move that call into `addImpl`? Looks like many/all uses need it.

File: sycl/source/detail/graph_impl.cpp [hunk 6]
```diff
@@ -726,6 +726,21 @@ node modifiable_command_graph::addImpl(std::function<void(handler &)> CGF,
...
 
+void modifiable_command_graph::addGraphLeafDependencies(node Node) {
+  // Find all exit nodes in the current graph and add them to the dependency
+  // vector
+  std::shared_ptr<detail::node_impl> DstImpl =
+      sycl::detail::getSyclObjImpl(Node);
+  graph_impl::WriteLock Lock(impl->MMutex);
+  for (auto &NodeImpl : impl->MNodeStorage) {
+    if (NodeImpl->MSuccessors.size() == 0) {
+      if (NodeImpl != DstImpl) {
```
Comment (aelovikov-intel) [inline sycl/source/detail/graph_impl.cpp:737]: Please re-write using early-continue style.

File: sycl/unittests/Extensions/CommandGraph.cpp [hunk 7]
```diff
@@ -1500,6 +1500,84 @@ TEST_F(CommandGraphTest, EnqueueMultipleBarrier) {
...
+  ASSERT_EQ(Node2Impl->MSuccessors[0].lock(), EmptyImpl);
+  auto Node3Impl = sycl::detail::getSyclObjImpl(Node3Graph);
+  ASSERT_EQ(Node3Impl->MSuccessors.size(), 1lu);
+
+  auto Node4Impl = sycl::detail::getSyclObjImpl(Node4Graph);
+  ASSERT_EQ(Node4Impl->MPredecessors.size(), 1lu);
+  ASSERT_EQ(Node4Impl->MSuccessors.size(), 1lu);
+  ASSERT_EQ(Node4Impl->MSuccessors[0].lock(), EmptyImpl);
+}
+
```
Comment (aelovikov-intel) [inline sycl/unittests/Extensions/CommandGraph.cpp:1580]: Would tests like below be meaningful here:

```
(1) (2)
 | \ |
 |  (E)
(4)  |
    (5)


(1) (2)
  \ /
  (E1)    (3)
    \     /
      (E2)
```

File: sycl/source/detail/graph_impl.cpp (hunk 6 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.cpp]: This coding style issue has been fixed.

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp (hunk 5 above)
Comment (mfrancepillois) [reply sycl/include/sycl/ext/oneapi/experimental/graph.hpp:198]: At the moment graph properties are handled at this level (graph front-end) and not in the graph implementation itself.
So, if you don't have a strong preference to see this call moved to addImpl, we probably prefer to keep it there.

Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/experimental/graph.hpp:198]: Looks like they all stay inside graph code and don't leak into regular SYCL RT code, so I'm fine.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by mfrancepillois** [comment]
@intel/llvm-gatekeepers could you please merge this PR

**Pull request merged**